applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+			NULL,
+			(struct params_wifi *)event->dyndata.data);
+		break;
+	default:
+		LOG_WRN("Unknown opt");
+		return;
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_module_state_event(eh)) {
+		const struct module_state_event *event;
+
+		event = cast_module_state_event(eh);
+
+		if (check_state(
+				event,
+				MODULE_ID(ble_state),
+				MODULE_STATE_READY)) {
+			static bool initialized;
+
+			__ASSERT_NO_MSG(!initialized);
+
+			mp_chmap_inst = chmap_filter_init();
+			__ASSERT_NO_MSG(mp_chmap_inst != NULL);
+
+			k_mutex_init(&m_data_access_mutex);
+			m_new_params = NULL;
+			m_new_blacklist = -1;
+
+			LOG_DBG("Chmap lib version: %s",
+				chmap_filter_version());
+			enable_qos_reporting();
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+			cdc_dev = device_get_binding(
+				CONFIG_USB_CDC_ACM_DEVICE_NAME "_0");
+			__ASSERT_NO_MSG(cdc_dev != NULL);
+# if CONFIG_UART_LINE_CTRL
+			cdc_dtr = 0;
+# else
+			cdc_dtr = 1;
+# endif
+#endif
+
+			initialized = true;
+
+			k_sem_give(&sem_initialized);
+		}
+
+		return false;
+	}
+
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			const struct config_event *event;
+
+			event = cast_config_event(eh);
+
+			if (!is_my_config_id(event->id)) {
+				return false;
+			}
+
+			handle_config_event(event);
+
+			return false;
+		}
+
+		if (is_config_fetch_request_event(eh)) {
+			const struct config_fetch_request_event *event =
+				cast_config_fetch_request_event(eh);
+
+			if (is_my_config_id(event->id)) {
+				struct config_fetch_event *fetch_event;
+
+				switch (OPT_FIELD_GET(event->id)) {
+				case QOS_OPT_BLACKLIST:
+					fetch_event =
+						fetch_event_qos_blacklist();
+					break;
+				case QOS_OPT_CHMAP:
+					fetch_event =
+						fetch_event_qos_channel_map();
+					break;
+				case QOS_OPT_PARAM_BLE:
+					fetch_event =
+						fetch_event_qos_ble_params();
+					break;
+				case QOS_OPT_PARAM_WIFI:
+					fetch_event =
+						fetch_event_qos_wifi_params();
+					break;
+				default:
+					LOG_WRN("Unknown opt");
+					return false;
+				}
+
+				fetch_event->id = event->id;
+				fetch_event->recipient = event->recipient;
+				fetch_event->channel_id = event->channel_id;
+
+				EVENT_SUBMIT(fetch_event);
+			}
+
+			return false;
+		}
+	}
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+	if (IS_ENABLED(CONFIG_DESKTOP_HID_MOUSE)) {
+		static s32_t m_hid_pkt_recv_count;
+		static s32_t m_usb_pkt_sent_count;
+		static u32_t m_cdc_notify_count;
+
+		/* Count number of HID packets received via BLE */
+		/* and number of HID packets sent via USB. */
+		/* Send stats printout via CDC every 100 packets */
+
+		if (is_hid_mouse_event(eh)) {
+			++m_hid_pkt_recv_count;
+			++m_cdc_notify_count;
+
+			if (m_cdc_notify_count == 100) {
+				hid_pkt_stats_print(
+					m_hid_pkt_recv_count,
+					m_usb_pkt_sent_count);
+				m_cdc_notify_count = 0;
+			}
+
+			return false;
+		}
+
+		if (is_hid_report_sent_event(eh)) {
```

Comment:
Yes, good point. I'm not sure what would be most useful: only look at mouse HID reports for the metrics printout, or count all HID reports. 
In this module the transmitted USB reports aren't as interesting as the BLE ones anyway, so these could be omitted altogether.

Reply:
Yes, received BLE reports seems to be more important here. And I think we may want to see mouse reports separated from the ones coming from keyboard (to properly calculate mouse report rate, which should be nearly constant when mouse is being moved all the time - it could be used as some sort of performance metrics).